Add Videoprism (#39895)
huggingface/transformers: docs/source/en/model_doc/videoprism.md

- For video classification tasks, use `VideoPrismForVideoClassification` which adds a classification head on top of the video encoder. `VideoPrismVisionConfig` must be used with this model.
- The vision encoder can be used standalone via `VideoPrismVisionModel` for extracting video features. `VideoPrismVisionConfig` must be used with this model.
- The default input resolution is 288x288 pixels with 16 frames per video clip for the base models and 8 frames for the large models. Set interpolate_pos_encoding=True to use the models with custom resolution and frames per clip.

This model was contributed by [MHRDYN7](https://github.com/MHRDYN7) and reviewed by [vasqu](https://github.com/vasqu) & [zucchini-nlp](https://github.com/zucchini-nlp).
The original code can be found [here](https://github.com/google-deepmind/videoprism).

## Usage example

The snippet below shows how to load the VideoPrismVisionModel for feature extraction using the `AutoModel` class.

```py
import torch
from transformers import AutoModel, AutoVideoProcessor

> processor = AutoVideoProcessor.from_pretrained("google/videoprism-base-f16r288")

Reviewer: we need to pass a revision here no?

<!--Copyright 2026 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->
*This model was released on 2025-06-03 and added to Hugging Face Transformers on 2026-05-19.*

<div style="float: right;">
    <div class="flex flex-wrap space-x-1">

>         <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">

Reviewer: ```suggestion
```
only a recent decision but we remove those now since we are torch only anyways

- The vision encoder can be used standalone via `VideoPrismVisionModel` for extracting video features. `VideoPrismVisionConfig` must be used with this model.
- The default input resolution is 288x288 pixels with 16 frames per video clip for the base models and 8 frames for the large models. Set interpolate_pos_encoding=True to use the models with custom resolution and frames per clip.

This model was contributed by [MHRDYN7](https://github.com/MHRDYN7) and reviewed by [qubvel](https://github.com/qubvel) & [zucchini-nlp](https://github.com/zucchini-nlp).
The original code can be found [here](https://github.com/google-deepmind/videoprism).

## Usage example

The snippet below shows how to load the VideoPrismVisionModel for feature extraction using the `AutoModel` class.

```py
import torch
from transformers import AutoModel, AutoVideoProcessor

processor = AutoVideoProcessor.from_pretrained("MHRDYN7/videoprism-base-f16r288")
model = AutoModel.from_pretrained(
    "MHRDYN7/videoprism-base-f16r288",

>     dtype=torch.float16,

Reviewer: ```suggestion
```
we use auto dtype by default so do we still need this?

    device_map="auto",
    attn_implementation="sdpa" # use "eager" to replicate the exact behavior as the original model
)

video_url = "https://huggingface.co/datasets/nateraw/kinetics-mini/resolve/main/val/archery/-Qz25rXdMjE_000014_000024.mp4"

# when do_sample_frames=True, 16/8 frames will be sampled by default depending on the checkpoint size base/large.
processed_video_inputs = processor(videos=[video_url], return_metadata=True, do_sample_frames=True)
video_metadata = processed_video_inputs["video_metadata"]
video_inputs = processed_video_inputs["pixel_values_videos"].to(model.device)
outputs = model(video_inputs)

# VideoPrism encoder outputs
encoder_outputs = outputs.last_hidden_state

```

> <Tip warning={true}>

Reviewer: I would move this a bit above, users usually dont read this far or at least are less likely to

Reviewer: We have this above now, so can be deleted here no?

Notes:

- VideoPrism uses a factorized spatio-temporal encoder architecture, processing videos through separate spatial and temporal transformers.
- The model supports video-text contrastive learning through `VideoPrismClipModel`, which combines a video encoder and a text encoder. `VideoPrismConfig` must be used with this model.
- For video classification tasks, use `VideoPrismForVideoClassification` which adds a classification head on top of the video encoder. `VideoPrismVisionConfig` must be used with this model.
- The vision encoder can be used standalone via `VideoPrismVisionModel` for extracting video features. `VideoPrismVisionConfig` must be used with this model.
- The default input resolution is 288x288 pixels with 16 frames per video clip for the base models and 8 frames for the large models. Set interpolate_pos_encoding=True to use the models with custom resolution and frames per clip.

This model was contributed by [MHRDYN7](https://github.com/MHRDYN7) and reviewed by [vasqu](https://github.com/vasqu) & [zucchini-nlp](https://github.com/zucchini-nlp).
The original code can be found [here](https://github.com/google-deepmind/videoprism).

> <Tip warning={true}>

Reviewer: Latest torchvision release now has lanczos interpolation

Author: yes, forgot to push that commit. Nevertheless, the processor does not exactly match with the mediapy version within some good tol (so I did not include equivalence test for this). I confirmed it with the torchvision maintainer last week, under that pr, and this is because the torchvision version is aligned with the faster Pillow-SIMD, rather than PIL. Though he said the differences are almost negligible, I found it high enough to not include the test. But still the processor is at least implemented correctly now.

Author: also, not sure if the CI testing deps are appropriate (torchvision==0.27 needed), but for me everything went well we with torch==2.12, torchcodec==0.13

Reviewer: Gotcha, would just adjust the messaging a bit then. Yea, CI is always a bit behind. We can use PIL backend for the tests? Or at least adjust the values a bit I think that's fair

Author: ok, for this I have also included the tensor chunks for bicubic (which is being used optionally if torchvision<0.27). Since both versions exist, once the CI torchvision version goes above 0.27, the bicubic ones can be deleted.

The video processor loaded via AutoProcessor is LlavaOnevisionVideoProcessor which is recomended for sampling frames exactly as in the original repository. However, please note that the [original processor](https://github.com/google-deepmind/videoprism/blob/main/videoprism/colabs/videoprism_video_encoder_demo.ipynb) uses Lanczos interpolation for resizing the frames, but that is not supported in pytorch yet and therefore LlavaOnevisionVideoProcessor uses Bicubic interpolation. 

</Tip>

## Usage example

The snippet below shows how to load the VideoPrismVisionModel for feature extraction using the `AutoModel` class.

```py
import torch
from transformers import AutoModel, AutoVideoProcessor

processor = AutoVideoProcessor.from_pretrained("google/videoprism-base-f16r288")
model = AutoModel.from_pretrained(
    "google/videoprism-base-f16r288",
    device_map="auto",

>     attn_implementation="eager" # use "flash_attention_2" for faster inference on supprted hardware

Reviewer: ```suggestion
    # use "flash_attention_2" for faster inference on supported hardware
    # attn_implementation="flash_attention_2" 
```

)

video_url = "https://huggingface.co/datasets/nateraw/kinetics-mini/resolve/main/val/archery/-Qz25rXdMjE_000014_000024.mp4"

vr = VideoDecoder(video_url)
frame_idx = np.arange(0, 64) # choosing some frames. here, you can define more complex sampling strategy
video = vr.get_frames_at(indices=frame_idx).data  # T x C x H x W

# automatically samples 16 frames by default for the base model
video = processor(video, return_tensors="pt").to(model.device)
outputs = model(**video)

# VideoPrism encoder outputs
encoder_outputs = outputs.last_hidden_state

```

You may also use the original video processing function provided in the VideoPrism repository examples. However, this will be slower than using torchcodec with VideoPrismVideoProcessor for large batches of videos.

```python
import numpy as np

def read_and_preprocess_video(
    filename: str, target_num_frames: int, target_frame_size: tuple[int, int]
):
    """Reads and preprocesses a video."""

    frames = mediapy.read_video(filename)

    # Sample to target number of frames.
    frame_indices = np.linspace(0, len(frames), num=target_num_frames, endpoint=False, dtype=np.int32)

>     frames = np.array([frames[i] for i in frame_indices])

Author: The default `sample_frames` in `BaseVideoProcessor`, samples the frames exactly in the same way as this `np.linspace` case above. The only thing that this original processor does differently is that mediapy.resize_video uses lanczos interpolation by default and that is not available in torch (so we use bicubic following the precedent of smolvlm). So, should I keep this info about the original processor in the doc?

>     frames = np.array([frames[i] for i in frame_indices])

Reviewer 2: Lte's just keep info about the resizing interpolation type as a TIP or Warning without the code

<!--Copyright 2025 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# VideoPrism

## Overview

The VideoPrism model was proposed in [<INSERT PAPER NAME HERE>](<INSERT PAPER LINK HERE>) by <INSERT AUTHORS HERE>.

> <INSERT SHORT SUMMARY HERE>

Reviewer 2: Let's fill-out the docs
